File: Data Structure I/2. String/P387.cpp
// Date: Sat, Oct 2, 2021
// 387. First Unique Character in a String (https://leetcode.com/problems/first-unique-character-in-a-string/submissions/)

// time complexity: O(n)
// space complexity: O(1)
int firstUniqChar(string s) {
	int hash[26] = {};
	// fill data to the hash table
	for (int i = 0; i < s.length(); i++)
		hash[s[i] - 'a'] += 1;
	// search for the target index
	for (int i = 0; i < s.length(); i++)
		if (hash[s[i] - 'a'] == 1)
			return i;
	return -1;
}

File: LeetCoding-Challenge/August-2022/P387.java
package leetcodingchallenge.august2022;

/**
 * Date: August 16, 2022 - Already solve it in C++ before
 * 387. First Unique Character in a String
 * https://leetcode.com/problems/first-unique-character-in-a-string/
 */

public class P387 {
    // time: O(n)
    // space: O(1)
    public int firstUniqChar(String s) {
        int[] counter = new int[26];
        for (char ch : s.toCharArray()) counter[ch - 'a']++;
        for (int i = 0; i < s.length(); i++) {
            if (counter[s.charAt(i) - 'a'] == 1) return i;
        }
        return -1;
    }
}
